sensor/common/compliance/nodescan_handler_impl.go

```diff
@@ -6,35 +6,51 @@ import (
 	"github.com/stackrox/rox/generated/storage"
 	"github.com/stackrox/rox/pkg/centralsensor"
 	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/sync"
 )
 
 type nodeScanHandlerImpl struct {
 	nodeScans <-chan *storage.NodeScanV2
-	toCentral chan *central.MsgFromSensor
+	toCentral <-chan *central.MsgFromSensor
+	// toCentralMux prevents the race condition between Start() [writer] and ResponsesC() [reader]
+	toCentralMux *sync.Mutex
 
-	stopC    concurrency.ErrorSignal
-	stoppedC concurrency.ErrorSignal
+	stopC concurrency.ErrorSignal
+	// runFinishedC is signaled when the goroutine inside of run() finishes
+	runFinishedC concurrency.ErrorSignal
+	startedC     concurrency.Signal
+}
+
+func (c *nodeScanHandlerImpl) Stopped() concurrency.ReadOnlyErrorSignal {
+	return &c.runFinishedC
 }
 
 func (c *nodeScanHandlerImpl) Capabilities() []centralsensor.SensorCapability {
 	return []centralsensor.SensorCapability{centralsensor.NodeScanningCap}
 }
 
 func (c *nodeScanHandlerImpl) ResponsesC() <-chan *central.MsgFromSensor {
+	c.toCentralMux.Lock()
+	defer c.toCentralMux.Unlock()
 	return c.toCentral
 }
 
 func (c *nodeScanHandlerImpl) Start() error {
-	// c.run closes chan toCentral on exit, so we do not want to close twice in case of a restart
+	c.toCentralMux.Lock()
+	defer c.toCentralMux.Unlock()
+	if c.startedC.IsDone() {
+		return errors.New("running handler must be stopped before it can be started again")
+	}
+	c.startedC.Signal()
 	if !c.stopC.IsDone() {
-		go c.run()
-		return nil
+		c.toCentral = c.run()
 	}
```

Comment:
I don't understand what does `!c.stopC.IsDone()` condition signify in the new logic.

If the component was never started before, the check passes and `run()` is called. No surprise but the check is not needed either.

If the component was started and currently running, the code wouldn't even get to this check.

If the component was started and `run()`-s inner goroutine completed, the code wouldn't even get to this check.

If the component was started and then shut down with the `Stop()` call, the condition will evaluate to `false` thus making `Start()` return without error but also without calling `run()`. I don't think that's a desired behavior.